### PR TITLE
attempts at PLEDGE VALIDATION: Validate Pregnancy#pledge_sent field in model #711

### PR DESCRIPTION
**Pull request by ashlynnpai**
Code is not working.

I tried a couple of things:

* Conditional validations, which we talked about. Here's an example from Railscasts: http://railscasts.com/episodes/41-conditional-validations. If the condition is true, the validation on pledge_sent gets triggered, otherwise it gets bypassed.

* Using the same method with before_update.

I'm out of ideas (I think) for keeping the logic in the model instead of blocking the update in the controller. Do you see a problem in my code?

### CODE REVIEW

**Comment by colinxfleming** [comment]
You're SUPER close on this! Like, mega close. Digging in now!

File: app/models/pregnancy.rb [hunk 1]
```diff
@@ -33,6 +33,15 @@ class Pregnancy
   # Validations
   validates :created_by,
             presence: true
+  before_update :check_other_validations?
+
+  def check_other_validations?
+    dcaf_soft_pledge.present? &&
+    Patient.first.name.present? &&
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: Swap these out for lowercase `patient` and remove the `first`s -- it'll be smart enough to pull the associated object thru activerecord.

File: app/models/pregnancy.rb [hunk 2]
```diff
@@ -33,6 +33,15 @@ class Pregnancy
...
+    Patient.first.appointment_date.present? &&
+    Patient.first.clinic.name.present?
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: Swap the `clinic.name` here out for `clinic_name` -- we haven't implemented Clinic objects yet, which is something I should have done a better job of documenting! Sorry, that's on me.

File: test/models/pregnancy_test.rb [hunk 3]
```diff
@@ -43,4 +43,23 @@ class PregnancyTest < ActiveSupport::TestCase
       assert @pregnancy.created_by
     end
   end
+
+  before do
+    @pregnancy.dcaf_soft_pledge = 500
+    @pt_1.appointment_date = DateTime.now + 14
+    @clinic = create :clinic
```
Comment (colinxfleming) [inline test/models/pregnancy_test.rb]: Same deal as above -- go ahead and set `@pt_1.clinic_name` to some string or another. Doesn't matter what just as long as it's not nil.

File: test/models/pregnancy_test.rb [hunk 4]
```diff
@@ -43,4 +43,23 @@ class PregnancyTest < ActiveSupport::TestCase
...
+  end
+
+  describe 'pledge_sent' do
+    it 'should update the pledge_sent field to true' do
+      @pregnancy.pledge_sent = true
+      assert_equal true, @pregnancy.pledge_sent
```
Comment (colinxfleming) [inline test/models/pregnancy_test.rb]: I think this might be a trouble spot -- because this is just checking value and not validation, I think this is always going to return true. 

Think of the common pattern in controllers -- 

```
if @patient.save
  # do successful thing
else
  # do failure thing
end
```

The actual `save` method is going to return true or false based on whether the object is valid and saves, which will then execute whichever part of the conditional. We could use `valid?` here too. But the value will hang out as true until you actually check for validity, which `save` does.

For example:

```
 irb(main):001:0> a = Patient.first
=> #<Patient _id: 58164096311c8852aca80137 ... >
irb(main):002:0> a.primary_phone = 'ashlynn is awesome' # set this to a string
=> "ashlynn is awesome"
irb(main):003:0> a.primary_phone # still a string
=> "ashlynn is awesome"
irb(main):004:0> a.valid? # check for validity, which will stop it 
=> false
irb(main):005:0> a.primary_phone # and now it's gone!
=> ""
```

I'd recommend changing this to `assert @pregnancy.valid?`. Let me know if this makes sense!

File: test/models/pregnancy_test.rb [hunk 5]
```diff
@@ -43,4 +43,23 @@ class PregnancyTest < ActiveSupport::TestCase
...
+    end
+
+    it 'should not update pledge_sent field to true if the DCAF soft pledge amount is nil' do
+      @pregnancy.dcaf_soft_pledge = nil
+      @pregnancy.pledge_sent = true
+      assert_equal nil, @pregnancy.pledge_sent
```
Comment (colinxfleming) [inline test/models/pregnancy_test.rb]: Similar deal here -- `pledge_sent` is going to always return true here because we haven't actually checked for validity yet. Swap the assertion here out to `refute @pregnancy.valid?`

File: app/models/pregnancy.rb [hunk 6]
```diff
@@ -33,6 +33,15 @@ class Pregnancy
   # Validations
   validates :created_by,
             presence: true
+  before_update :check_other_validations?
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: Ahh you are so close! We should try and do this with validations rather than hooks if we can. Here's a suggested potential way that use the `check_other_validations?` method you wrote:

* Exclusions: http://guides.rubyonrails.org/active_record_validations.html#exclusion
Something like `exclusion: in: [ true ], unless: check_other_validations?` (In plain english: 'Value cannot be true, unless check other validations is also true.`

**Comment by ashlynnpai** [comment]
I tried to get the integration test to pass by supplying additional data. The test still failed and the data seemed to intermittently persist to other tests, so just commented out.

**Comment by ashlynnpai** [comment]
I should refactor the test data in the model tests when we figure out how to get this working.

File: test/integration/submit_pledge_test.rb [hunk 7]
```diff
@@ -6,6 +6,9 @@ class SubmitPledgeTest < ActionDispatch::IntegrationTest
     @user = create :user
     @patient = create :patient
```
Comment (colinxfleming) [inline test/integration/submit_pledge_test.rb]: FactoryGirl lets you create stuff and assign attributes in one line -- if you do `@patient = create :patient, clinic_name: 'Nice Clinic', appointment_date: DateTime.now + 14` and `@pregnancy = create :pregnancy, patient: @patient, dcaf_soft_pledge: 500` it'll create the objects and set those attributes while it does it. 

I got the test to pass using this! Maybe it wasn't saving as written?

Comment (ashlynnpai) [inline test/integration/submit_pledge_test.rb]: I thought you could set attributes after creation. I'll take another look tonight but glad you got it to pass.

Comment (ashlynnpai) [inline test/integration/submit_pledge_test.rb]: I see, there was a submit at the end. Thanks for catching that!!

**Comment by ashlynnpai** [comment]
Thanks man!! Let me know if this works for you.

File: test/models/pregnancy_test.rb [hunk 8]
```diff
@@ -43,4 +43,35 @@ class PregnancyTest < ActiveSupport::TestCase
...
+    @pregnancy.dcaf_soft_pledge = 500
+    @pt_1.clinic_name = 'Nice Clinic'
+    @pt_1.appointment_date = DateTime.now + 14    
+  end
+
+  describe 'pledge_sent validation' do
+    it 'should not validate pledge_sent if the DCAF soft pledge field is blank' do
+      @pregnancy.dcaf_soft_pledge = nil
+      @pregnancy.pledge_sent = true
+      refute @pregnancy.valid?
```
Comment (colinxfleming) [inline test/models/pregnancy_test.rb]: Oh man this is great, A++++ work. Love how readable and clear this is. 

Can you confirm in these sets of tests that the right error message comes back too? I think that'll keep the test coverage at 100%, and we've got some modal stuff coming up that will consume that.

File: app/models/pregnancy.rb [hunk 9]
```diff
@@ -50,4 +51,22 @@ class Pregnancy
...
   # end
+  
+  private
+  
+  def updating_pledge_sent?
+    pledge_sent == true
+  end
+
+  def check_other_validations?
+    if dcaf_soft_pledge.blank?
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: If these aren't set, we want them to all spit out, yeah? What are your thoughts on turning these into a bunch of one-liners? e.g.:

```
errors.add(:pledge_sent, 'DCAF soft pledge field cannot be blank') if dcaf_soft_pledge.blank?
# ... and so on
```

File: app/models/pregnancy.rb [hunk 10]
```diff
@@ -33,6 +33,7 @@ class Pregnancy
   # Validations
   validates :created_by,
             presence: true
+  validate :pledge_sent, :check_other_validations?, if: :updating_pledge_sent?
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: I like this a lot -- validating only when changing it to true is a really creative and clever implementation! Nicely done 💯

File: app/models/pregnancy.rb [hunk 11]
```diff
@@ -50,4 +51,22 @@ class Pregnancy
...
   #   false
   # end
+  
+  private
+  
+  def updating_pledge_sent?
+    pledge_sent == true
+  end
+
+  def check_other_validations?
```
Comment (colinxfleming) [inline app/models/pregnancy.rb]: What are your thoughts about renaming this method? Usually having `?` at the end of a method indicates it either returns true or false, and this isn't really doing this in the current (very good and useful!) implementation. Instead, it's confirming that necessary information is saved so it can mark the pledge sent. Thoughts?

File: app/models/pregnancy.rb (hunk 9 above)
Comment (ashlynnpai) [inline app/models/pregnancy.rb]: Sure. Will work on it this week.

**Comment by ashlynnpai** [comment]
I seem to be having problems again with the test data not resetting.

**Comment by colinxfleming** [comment]
@ashlynnpai are you still hitting that snag? This looks pretty good to me, and the test failure is from one of our scrub tests.

**Comment by colinxfleming** [comment]
This looks to me like we can merge it if you're feeling good about it @ashlynnpai ! Let me know!

**Comment by ashlynnpai** [comment]
Yes, I ran the test suite three times. Looks good.

**Comment by colinxfleming** [comment]
@ashlynnpai you are number one! Thank you so much for pushing this over the finish line and staying on this!

**Comment by colinxfleming** [comment]
Fixes #711 🎉
